// gulp-less-changed/release/data-uri-visitor.js
"use strict";
var path = require("path");
var dataUriVisitor;
(function (dataUriVisitor) {
    var DataUriVisitor = (function () {
        function DataUriVisitor(less) {
            this.isReplacing = false;
            this.isPreEvalVisitor = true;
            this._imports = [];
            this._visitor = new less.visitors.Visitor(this);
        }
        DataUriVisitor.prototype.run = function (root) {
            return this._visitor.visit(root);
        };
        DataUriVisitor.prototype.tryGetImportedFileName = function (ruleNode) {
            var fileName;
            if (ruleNode.args.length === 2) {
                fileName = ruleNode.args[1];
            }
            else {
                fileName = ruleNode.args[0];
            }
            if (!fileName.value || /@/.test(fileName.value)) {
                return null;
            }
            return fileName.value;
        };
        DataUriVisitor.prototype.getImportInfo = function (ruleNode) {
            if (ruleNode.name !== 'data-uri' ||
                ruleNode.args.length === 0) {
                return { ruleNode: ruleNode };
            }
            var importedFile = this.tryGetImportedFileName(ruleNode);
            if (!importedFile) {
                return { ruleNode: ruleNode };
            }
            var fileInfo = typeof ruleNode.fileInfo === 'function' ? ruleNode.fileInfo() : ruleNode.currentFileInfo;
            var entryPath = fileInfo.entryPath;
            return { ruleNode: ruleNode, importedFile: importedFile, entryPath: entryPath };
        };
        DataUriVisitor.prototype.visitCall = function (callNode, visitArgs) {
            var _a = this.getImportInfo(callNode), ruleNode = _a.ruleNode, importedFile = _a.importedFile, entryPath = _a.entryPath;
            if (!importedFile) {
                return ruleNode;
            }
            this._imports.push({ directory: entryPath ? path.normalize(entryPath) : '', relativePath: importedFile });
            return ruleNode;
        };
        Object.defineProperty(DataUriVisitor.prototype, "imports", {
            get: function () {
                return this._imports;
            },
            enumerable: true,
            configurable: true
        });
        return DataUriVisitor;
    }());
    dataUriVisitor.DataUriVisitor = DataUriVisitor;
})(dataUriVisitor || (dataUriVisitor = {}));
module.exports = dataUriVisitor;
